### PR TITLE
Send MAC commands confirmed in class C

### LINKED ISSUE #317: NS keeps scheduling class C downlink every 15 seconds
**Summary:** 

NS warns that session is empty after OTAA join.

**Steps to Reproduce:** 

1. Checkout current tip `03696ccb`
2. Start stack; `ttn-lw-stack start`
3. Create device; `ttn-lw-cli dev create test-app simulation3 --lorawan-version 1.0.2 --lorawan-phy-version 1.0.2-b --frequency-plan-id EU_863_870 --join-eui 4200000000000000 --dev-eui 4200000000000002 --root-keys.app-key.key 42000000000000000000000000000000`
4. Run the simulator (see https://github.com/johanstokking/mbed-os-lorawan-simulation)
5. Enable class C; `ttn-lw-cli dev set test-app simulation3 --mac-state.device-class=CLASS_C`
6. Schedule downlink; `mosquitto_pub -h localhost -u test-app -P $APP_API_KEY -d -t 'v3/test-app/devices/simulation3/down/push' -m '{"downlinks":[{"f_port":15,"frm_payload":"vu8="}]}'`

**What do you see now?**

Getting logs every 15 seconds;

```
 DEBUG Processing downlink task...              delay=76.622ms device_uid=test-app.simulation3 start_at=2019-03-19 16:39:53.92445 +0000 UTC
 DEBUG Generated downlink                       ack=false delay=76.622ms device_uid=test-app.simulation3 f_cnt=406 f_pending=true f_port=0 m_type=UNCONFIRMED_DOWN mac_count=7 payload_length=46 start_at=2019-03-19 16:39:53.92445 +0000 UTC
 DEBUG Scheduling downlink...                   attempt_rx1=false attempt_rx2=true delay=76.622ms device_class=CLASS_C device_uid=test-app.simulation3 downlink_type=data path_count=1 start_at=2019-03-19 16:39:53.92445 +0000 UTC
 DEBUG Scheduling downlink                      class=CLASS_C gateway_eui=ACDE480000001122 gateway_uid=simulator namespace=gatewayserver/io/udp
 DEBUG Scheduled downlink                       class=CLASS_C data_rate=0 duration=2.301952s frequency=869525000 gateway_eui=ACDE480000001122 gateway_uid=simulator namespace=gatewayserver/io/udp rx_window=2 starts=74622464000
 DEBUG Writing downlink message                 gateway_eui=ACDE480000001122 gateway_uid=simulator namespace=gatewayserver/io/udp remote_addr=127.0.0.1:65454
  INFO Finished unary call                      duration=1.019936ms grpc_method=ScheduleDownlink grpc_service=ttn.lorawan.v3.NsGs namespace=grpc request_id=01D6BD5NWZS7BBMBVEJHRSVR8J
 DEBUG Finished unary call                      duration=1.413197ms grpc_method=ScheduleDownlink grpc_service=ttn.lorawan.v3.NsGs namespace=grpc
 DEBUG Scheduled downlink                       attempt_rx1=false attempt_rx2=true delay=0s device_class=CLASS_C device_uid=test-app.simulation3 downlink_type=data start_at=2019-03-19 16:39:53.92445 +0000 UTC
 DEBUG Recalculating downlink queue             application_uid=test-app count=1 device_uid=test-app.simulation3 network_server=Johans-MacBook-Pro.local next_a_f_cnt_down=407
 DEBUG Run database query                       duration=2ms grpc_method=DownlinkQueueReplace grpc_service=ttn.lorawan.v3.AsNs namespace=db query=SELECT * FROM "api_keys"  WHERE ("api_keys"."api_key_id" = $1) ORDER BY "api_keys"."id" ASC LIMIT 1 request_id=01D6BD5NX2AWN25CWBM6RSX0RC rows=1 source=api_key_store.go:71 values=[ILZVXXQBP4QVCPYPXIYVC6CPKBSAUOX65BW2V7Y]
 DEBUG Adding device to downlink schedule...    delay=76.622ms device_class=CLASS_C device_uid=test-app.simulation3 schedule_at=2019-03-19 16:40:09.015039 +0000 UTC start_at=2019-03-19 16:39:53.92445 +0000 UTC
 DEBUG Run database query                       duration=3ms grpc_method=DownlinkQueueReplace grpc_service=ttn.lorawan.v3.AsNs namespace=db query=SELECT id as uuid, application_id as friendly_id FROM "applications"  WHERE (id in ($1)) request_id=01D6BD5NX2AWN25CWBM6RSX0RC rows=1 source=membership.go:90 values=[ffe318d4-5750-44ec-903e-2068c42abe21]
 DEBUG Replaced application downlink queue      device_class=CLASS_C device_uid=test-app.simulation3 grpc_method=DownlinkQueueReplace grpc_service=ttn.lorawan.v3.AsNs namespace=grpc queue_length=1 request_id=01D6BD5NX2AWN25CWBM6RSX0RC
 DEBUG Adding downlink task...                  device_class=CLASS_C device_uid=test-app.simulation3 grpc_method=DownlinkQueueReplace grpc_service=ttn.lorawan.v3.AsNs namespace=grpc queue_length=1 request_id=01D6BD5NX2AWN25CWBM6RSX0RC
  INFO Finished unary call                      duration=29.934419ms grpc_method=DownlinkQueueReplace grpc_service=ttn.lorawan.v3.AsNs namespace=grpc request_id=01D6BD5NX2AWN25CWBM6RSX0RC
 DEBUG Finished unary call                      duration=30.471478ms grpc_method=DownlinkQueueReplace grpc_service=ttn.lorawan.v3.AsNs namespace=grpc
```

Note that it seems that this is AS initiated, but there is no AS task that runs on 15 seconds intervals. The obvious thing I find in the code base is the `DefaultClassCTimeout` being 15 seconds; an NS thing.

The item is being pushed back in the queue continuously.

```
$ ttn-lw-cli dev downlink list test-app simulation3
 DEBUG Using access token (valid until 6:17PM)
 DEBUG pickfirstBalancer: HandleSubConnStateChange: 0xc000458b60, CONNECTING
 DEBUG pickfirstBalancer: HandleSubConnStateChange: 0xc000458b60, READY
 DEBUG Finished unary call                      duration=69.153011ms grpc_method=DownlinkQueueList grpc_service=ttn.lorawan.v3.AppAs namespace=grpc
[{
  "session_key_id": "AWmWxBSWcVh7EZSRPpRgxA==",
  "f_port": 15,
  "f_cnt": 378,
  "frm_payload": "vu8=",
  "correlation_ids": [
    "as:conn:01D6BCAEK3CBT42ME6TND67MAT",
    "as:downlink:01D6BCAEK32G9B1Q7X64PFTD2D"
  ]
}]
```

See the high `f_cnt`.

Is NS engaging the replacement by AS via non-acked downlink? Note that this downlink is not confirmed.

**What do you want to see instead?**

Fire and forget for non-confirmed downlink.

**How do you propose to implement this?**

Not sure.

**What can you do yourself and what do you need help with?**

Review.

### LINKED ISSUE #406: Send request MAC commands confirmed when in class C
**Summary:** 

Send request MAC commands confirmed when in class C. This resolves #317. 

**Why do we need this?**

In LoRaWAN 1.1, a paragraph has been added regarding time-outs of downlink messages that need a response (i.e. confirmed and MAC commands):

>In case a message is received by a device in Class C mode requiring an uplink transmission (DL MAC command request or DL message in confirmed mode), the device SHALL answer within a time period known by both the end-device and the Network Server (out-of-band provisioning information).
>Before this timeout expires, the network SHALL not send any new confirmed message or MAC command to the device. Once this timeout expires or after reception of any uplink message, the network is allowed to send a new DL message.

This is implemented in V3 as well, using a 15 second timeout. This implies that 1.1 devices are sending immediate uplinks on request MAC commands. However, before 1.1, this does not happen, and the answer MAC command comes on the first uplink, which can be anytime in the future. However, per the V3 implementation, all other class C downlink is blocked.

**What is already there? What do you see now?**

Currently every 15 seconds the Network Server sends `DevStatusReq` , while the device with LoRaWAN 1.0.x only answer on its first uplink.

**What is missing? What do you want to see?**

Send downlink carrying request MAC commands as confirmed when the device is in class C mode.

**How do you propose to implement this?**

Not sure yet.

**Can you do this yourself and submit a Pull Request?**

Yes.

### PR DESCRIPTION
**Pull request by johanstokking**
**Summary:**

Closes #317 
Closes #406 

**Changes:**

- Send MAC commands confirmed in class C
- Minor cleanup

### CODE REVIEW

File: pkg/networkserver/downlink.go [hunk 1]
```diff
@@ -144,15 +144,20 @@ func (ns *NetworkServer) generateDownlink(ctx context.Context, dev *ttnpb.EndDev
...
+		logger := logger.WithField("cid", cmd.CID)
+		logger.Debug("Adding MAC command to buffer...")
 		var err error
-		logger.WithField("cid", cmd.CID).Debug("Adding MAC command to buffer...")
 		cmdBuf, err = spec.AppendDownlink(cmdBuf, *cmd)
-
 		if err != nil {
 			return nil, errEncodeMAC.WithCause(err)
 		}
+		if mType == ttnpb.MType_UNCONFIRMED_DOWN && spec[cmd.CID].ExpectAnswer && dev.MACState.DeviceClass == ttnpb.CLASS_C {
```
Comment (htdvisser) [inline pkg/networkserver/downlink.go]: Shouldn't we also do this for class B?

Comment (johanstokking) [reply pkg/networkserver/downlink.go]: The reason to make the downlink confirmed is for the device to respond within the class C timeout (V3 default 15 seconds). If it responds later, NS keeps sending MAC requests.

In class B, we need to make them confirmed as well, if the device does not respond before the next ping slot. LoRaWAN 1.0.3 does not specify answering class B frames; it only states that they're handled like class A, and in class A, the answer may come later than the next ping slot.

So I'm fine confirmed downlink of MAC commands where expect an answer in class B as well, but this is not based on device behavior but rather what's not specified.

Comment (htdvisser) [reply pkg/networkserver/downlink.go]: Is this something we need to raise in the TC as well?

Comment (johanstokking) [reply pkg/networkserver/downlink.go]: Added reference in #19 so that we don't forget. If I have time I'll raise a CR for Berlin AMM.

**Pull request merged**